staging_vespalib/src/vespa/vespalib/metrics/simple_metrics_manager.cpp

```diff
@@ -0,0 +1,198 @@
+{
+    _stopFlag = true;
+    _collectorThread.join();
+}
+
+
+std::shared_ptr<MetricsManager>
+SimpleMetricsManager::create(const SimpleManagerConfig &config)
+{
+    return std::shared_ptr<MetricsManager>(
+        new SimpleMetricsManager(config));
+}
+
+Counter
+SimpleMetricsManager::counter(const vespalib::string &name)
+{
+    size_t id = _metricNames.resolve(name);
+    _metricTypes.check(id, name, MetricTypes::COUNTER);
+    LOG(debug, "metric name %s -> %zd", name.c_str(), id);
+    return Counter(shared_from_this(), MetricIdentifier(id));
+}
+
+Gauge
+SimpleMetricsManager::gauge(const vespalib::string &name)
+{
+    size_t id = _metricNames.resolve(name);
+    _metricTypes.check(id, name, MetricTypes::GAUGE);
+    LOG(debug, "metric name %s -> %zd", name.c_str(), id);
+    return Gauge(shared_from_this(), MetricIdentifier(id));
+}
+
+Bucket
+SimpleMetricsManager::mergeBuckets()
+{
+    Guard bucketsGuard(_bucketsLock);
+    if (_buckets.size() > 0) {
+        InternalTimeStamp startTime = _buckets[_firstBucket].startTime;
+        Bucket merger(startTime, startTime);
+        for (size_t i = 0; i < _buckets.size(); i++) {
+            size_t off = (_firstBucket + i) % _buckets.size();
+            merger.merge(_buckets[off]);
+        }
+        return merger;
+    }
+    // no data
+    return Bucket(_startTime, _curTime);
+}
+
+Snapshot
+SimpleMetricsManager::snapshot()
+{
+    Bucket merged = mergeBuckets();
+    std::vector<PointSnapshot> points;
+
+    std::chrono::microseconds s = since_epoch(merged.startTime);
+    std::chrono::microseconds e = since_epoch(merged.endTime);
+    const double micro = 0.000001;
+    Snapshot snap(s.count() * micro, e.count() * micro);
+    {
+        Guard guard(_pointMaps.lock);
+        for (auto entry : _pointMaps.vec) {
+             const PointMapBacking &map = entry->first.backing();
+             PointSnapshot point;
+             for (const PointMapBacking::value_type &kv : map) {
+                 point.dimensions.emplace_back(nameFor(kv.first), valueFor(kv.second));
+             }
+             snap.add(point);
+        }
+    }
+    for (const CounterAggregator& entry : merged.counters) {
+        size_t ni = entry.idx.name_idx;
+        size_t pi = entry.idx.point_idx;
+        const vespalib::string &name = _metricNames.lookup(ni);
+        CounterSnapshot val(name, snap.points()[pi], entry);
+        snap.add(val);
+    }
+    for (const GaugeAggregator& entry : merged.gauges) {
+        size_t ni = entry.idx.name_idx;
+        size_t pi = entry.idx.point_idx;
+        const vespalib::string &name = _metricNames.lookup(ni);
+        GaugeSnapshot val(name, snap.points()[pi], entry);
+        snap.add(val);
+    }
+    return snap;
+}
+
+void
+SimpleMetricsManager::doCollectLoop(SimpleMetricsManager *me)
+{
+    const std::chrono::milliseconds jiffy{20};
+    const std::chrono::seconds oneSec{1};
+    while (!me->_stopFlag) {
+        std::this_thread::sleep_for(jiffy);
```

Comment:
consider waiting a full second at a condition instead (and signal it in destructor)